Fix oadp-dev CI for velero 1.17, no restic 1.6+lint

kaovilai opened the pull request:
- **Remove support for Restic in Data Protection Application and update related tests**
- **Refactor usage of pointer utilities to use the new ptr package and improve error messages in various controllers and tests**

## Why the changes were made

<!-- Explain why this PR is important, what problems it fixes, link related issues -->

## How to test the changes made

<!-- Explain how the changes introduced by this PR can be tested and verified, with commands and pictures -->

kaovilai committed:
Remove support for Restic in Data Protection Application and update related tests

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai committed:
Refactor usage of pointer utilities to use the new ptr package and improve error messages in various controllers and tests

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai committed:
`make bundle`

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

shubham-pampattiwar commented on api/v1alpha1/dataprotectionapplication_types.go:
```diff
@@ -452,6 +451,8 @@ type NodeAgentConfig struct {
+	// TODO: remove restic from below uploaderType enum?
+
 	// The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'
 	// +kubebuilder:validation:Enum=restic;kopia
```
Shouldn't we be removing the uploader type option as "restic" for install ? https://github.com/vmware-tanzu/velero/releases/tag/v1.17.0

sseago replied:
@shubham-pampattiwar I think we don't want to remove it as a valid CRD value -- that would prevent upgrade from working, right? It's valid in the CRD, but will fail reconcile.

kaovilai replied:
removing enum from validation is a breaking change which would fail https://operator-framework.github.io/operator-controller/concepts/crd-upgrade-safety/

I think we wanna do soft validation only. perhaps remove from CRD once we have a version where restic is not allowed to reconcile.

sseago replied:
@kaovilai if we're going to keep it in the enum (to avoid making incompatible CRD changes) then we may want to remove the TODO here as well. If/when we move to v2 CRD, then we could remove it (and various other things)

sseago commented on bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml:
```diff
@@ -2655,7 +2655,6 @@ spec:
                       - openshiftPluginImageFqin
                       - azurePluginImageFqin
                       - gcpPluginImageFqin
-                      - resticRestoreImageFqin
```
This is less about restic and more about the fact that we no longer use a separate restore-helper image. I don't know that we ever renamed it from restic-restore-helper. In any case, since oadp 1.5 we use the velero image for this, so I don't think we need the above anymore. It should be safe to remove.

sseago commented on internal/controller/validator.go:
```diff
@@ -116,15 +114,7 @@ func (r *DataProtectionApplicationReconciler) ValidateDataProtectionCR(log logr.
 
 	// DEPRECATIONS -----------------------------------------------------------
```
Should we remove the // DEPRECATIONS comment now, since it's no longer deprecated functionality, it's completely removed?

kaovilai replied:
reworded. its about removed features now.

shubham-pampattiwar reviewed (commented):
/lgtm

kaovilai commented:
/test 4.20-e2e-test-aws

kaovilai commented:
/retest

kaovilai commented:
```
  [FAIL] Backup and restore tests Backup and restore applications [It] Mongo application KOPIA
```

kopia isn't working

```
time="2025-09-26T19:16:19Z" level=error msg="Error backing up item" backup=openshift-adp/mongo-kopia-e2e-2891e3c8-9b0d-11f0-87a9-0a580a831b6b error="unable to create backup repository resource: BackupRepository.velero.io \"mongo-persistent-ts-velero-test-1-kopia\" is invalid: spec.resticIdentifier: Required value" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/repository/ensurer.go:112" error.function="github.com/vmware-tanzu/velero/pkg/repository.(*Ensurer).createBackupRepositoryAndWait" logSource="pkg/backup/backup.go:863" name=mongo-746cc748bf-msf9f
```

kaovilai commented:
need crd update

mpryc committed:
Update CRDs from  velero:main (99f12b8)

Updated CRDs from Velero oadp-dev.

Signed-off-by: Michal Pryc <mpryc@redhat.com>

kaovilai commented:
CRD from https://github.com/openshift/velero/pull/452 added

mpryc committed:
UPSTREAM: <drop>: Updating go modules

Signed-off-by: Michal Pryc <mpryc@redhat.com>

kaovilai committed:
fix `go mod/vet ./...` && `make bundle`

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai commented on internal/controller/nodeagent.go:
```diff
@@ -253,7 +253,9 @@ func (r *DataProtectionApplicationReconciler) ReconcileNodeAgentDaemonset(log lo
 			veleroAffinityStruct := make([]*kube.LoadAffinity, len(dpa.Spec.Configuration.NodeAgent.NodeAgentConfigMapSettings.LoadAffinityConfig))
 
 			for i, aff := range dpa.Spec.Configuration.NodeAgent.NodeAgentConfigMapSettings.LoadAffinityConfig {
-				veleroAffinityStruct[i] = (*kube.LoadAffinity)(aff)
+				veleroAffinityStruct[i] = &kube.LoadAffinity{
+					NodeSelector: aff.NodeSelector,
+				}
```
Add StorageClass to dpa.LoadAffinity type
velero added StorageClass to affinity so follow up to add storageclass to our [types](https://github.com/kaovilai/oadp-operator/blob/dde9fe2da2f6e689d25e513f7d9b3c2368efe4aa/api/v1alpha1/dataprotectionapplication_types.go#L399-L403)
from velero pkg/util/kube/pod.go
```go
type LoadAffinity struct {
	// NodeSelector specifies the label selector to match nodes
	NodeSelector metav1.LabelSelector `json:"nodeSelector"`

	// StorageClass specifies the VGDPs the LoadAffinity applied to. If the StorageClass doesn't have value, it applies to all. If not, it applies to only the VGDPs that use this StorageClass.
	StorageClass string `json:"storageClass"`
}
```

kaovilai commented:
unit test blocked by https://github.com/openshift/release/pull/69860

kaovilai commented:
/test unit-test

kaovilai committed:
`make generate`

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai committed:
Implement manual DeepCopy for NodeAgentConfigMapSettings and remove autogenerated version

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai commented:
also need to bump other jobs

kaovilai commented:
now blocked by https://github.com/openshift/release/pull/69867

sseago commented:
/lgtm

kaovilai commented:
```
  [FAILED] Expected
      <[]string | len:1, cap:1>: [
          "time=\"2025-10-01T00:13:40Z\" level=error msg=\"pod volume backup failed: data path backup failed: Failed to run data path service for PVB mongo-kopia-e2e-532e201c-9e5b-11f0-b7a6-0a580a83af0c-vmdln: Data path for PVB mongo-kopia-e2e-532e201c-9e5b-11f0-b7a6-0a580a83af0c-vmdln failed: Failed to run kopia backup: Failed to upload the kopia snapshot for si default@default:snapshot-pod-volume/kopia/mongo-persistent/mongo-746cc748bf-hmg2s/mongo: permission denied\" backup=openshift-adp/mongo-kopia-e2e-532e201c-9e5b-11f0-b7a6-0a580a83af0c logSource=\"pkg/podvolume/backupper.go:430\"",
      ]
  to equal
      <[]string | len:0, cap:0>: []
  In [It] at: /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_suite_test.go:192 @ 10/01/25 00:13:46.755
  < Exit [It] Mongo application KOPIA - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_suite_test.go:368 @ 10/01/25 00:13:46.755 (1m25.375s)
  > Enter [AfterEach] Backup and restore tests - /go/src/github.com/openshift/oadp-operator/tests/e2e/backup_restore_suite_test.go:300 @ 10/01/25 00:13:46.755
```

do we need some other fix here? @sseago

kaovilai commented:
might need https://github.com/vmware-tanzu/velero/pull/9300/files#diff-fd7c63f8e6ecf9a057d87984b1a410cc135dcb14e3d0d111dcac93e5b77e7c33

kaovilai committed:
Add PrivilegedFsBackup field to NodeAgentConfigMapSettings

  Add new optional PrivilegedFsBackup field to control whether fs-backup
  pods run in privileged security context. Defaults to true for OpenShift
  compatibility.

  Also update mongo-persistent test applications to use IfNotPresent
  imagePullPolicy.

kaovilai commented:
time=\"2025-10-01T04:57:07Z\" level=error msg=\"pod volume backup failed: data path backup failed: Failed to run data path service for PVB mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237-xxvpt: Data path for PVB mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237-xxvpt failed: Failed to run kopia backup: Failed to upload the kopia snapshot for si default@default:snapshot-pod-volume/kopia/mongo-persistent/mongo-8c6bcdbff-k7nlg/mongo: permission denied\" backup=openshift-adp/mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237 logSource=\"pkg/podvolume/backupper.go:430\""

weshayutin commented:
> time="2025-10-01T04:57:07Z" level=error msg="pod volume backup failed: data path backup failed: Failed to run data path service for PVB mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237-xxvpt: Data path for PVB mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237-xxvpt failed: Failed to run kopia backup: Failed to upload the kopia snapshot for si default@default:snapshot-pod-volume/kopia/mongo-persistent/mongo-8c6bcdbff-k7nlg/mongo: permission denied" backup=openshift-adp/mongo-kopia-e2e-ebedad2a-9e82-11f0-843b-0a580a81b237 logSource="pkg/podvolume/backupper.go:430""

Isn't that the issue that @sseago fixed upstream that QE was also hitting on 1.17?

sseago commented:
@weshayutin Yes. That will fail until my PR merges -- I was waiting to submit that until this one merges. So we can override on fs-backup failures, right?

weshayutin commented:
> @weshayutin Yes. That will fail until my PR merges -- I was waiting to submit that until this one merges. So we can override on fs-backup failures, right?

yes sir. .thanks all!
